### PR TITLE
  Add/Remove contract spec from scope spec.

### PR DESCRIPTION
**Pull request by channa-figure**
## Description

Add/Remove contract spec from scope spec.

closes: #302 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/provenance-io/provenance/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Wrote unit and integration [tests](https://github.com/provenance-io/provenance/blob/main/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [x] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [x] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Review `Codecov Report` in the comment section below once CI passes

### CODE REVIEW

File: x/metadata/client/cli/tx.go [hunk 1]
```diff
@@ -489,6 +492,46 @@ icon-url           - address to a image to be used as an icon (optional, can onl
...
+		Short:   "Add an existing contract specification to a scope specification",
+		Aliases: []string{"acstss"},
+		Args:    cobra.ExactArgs(2),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			clientCtx, err := client.GetClientTxContext(cmd)
+			if err != nil {
+				return err
+			}
+			contractSpecID, err := types.MetadataAddressFromBech32(args[0])
+			if err != nil {
```
Comment (iramiller) [inline x/metadata/client/cli/tx.go:508]:
```suggestion
			if err != nil || !contractSpecID.IsContractSpecificationAddress(){
```
Should verify parameter is a contract spec address. note: will need to handle/create an error for this to use with following line

File: x/metadata/client/cli/tx.go [hunk 2]
```diff
@@ -489,6 +492,46 @@ icon-url           - address to a image to be used as an icon (optional, can onl
...
+				return fmt.Errorf("invalid contract specification id : %v", err)
+			}
+			scopeSpecID, err := types.MetadataAddressFromBech32(args[1])
+			if err != nil {
```
Comment (iramiller) [inline x/metadata/client/cli/tx.go:512]:
```suggestion
			if err != nil || !scopeSpecID.IsScopeSpecificationAddress() {
```
verify address type is scope specification. note: will need to handle/create an error for this to use with following line

File: x/metadata/client/cli/tx.go [hunk 3]
```diff
@@ -1033,6 +1076,46 @@ func RemoveContractSpecificationCmd() *cobra.Command {
...
+		Short:   "Remove an existing contract specification from a scope specification",
+		Aliases: []string{"rcsfss"},
+		Args:    cobra.ExactArgs(2),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			clientCtx, err := client.GetClientTxContext(cmd)
+			if err != nil {
+				return err
+			}
+			contractSpecID, err := types.MetadataAddressFromBech32(args[0])
+			if err != nil {
```
Comment (iramiller) [inline x/metadata/client/cli/tx.go:1092]:
```suggestion
			if err != nil  || !contractSpecID.IsContractSpecificationAddress() {
```
verify address type... note: will need to handle/create an error for this to use with following line

File: x/metadata/client/cli/tx.go [hunk 4]
```diff
@@ -1033,6 +1076,46 @@ func RemoveContractSpecificationCmd() *cobra.Command {
...
+				return fmt.Errorf("invalid contract specification id : %v", err)
+			}
+			scopeSpecID, err := types.MetadataAddressFromBech32(args[1])
+			if err != nil {
```
Comment (iramiller) [inline x/metadata/client/cli/tx.go:1096]:
```suggestion
			if err != nil || !scopeSpecID.IsScopeSpecificationAddress()  {
```
verify type is scope specification adddress, note: will need to handle/create an error for this to use with following line

File: x/metadata/keeper/msg_server.go [hunk 5]
```diff
@@ -368,6 +368,83 @@ func (k msgServer) DeleteContractSpecification(
...
+	goCtx context.Context,
+	msg *types.MsgAddContractSpecToScopeSpecRequest,
+) (*types.MsgAddContractSpecToScopeSpecResponse, error) {
+	defer telemetry.MeasureSince(time.Now(), types.ModuleName, "tx", "AddContractSpecToScopeSpec")
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	contractSpec, found := k.GetContractSpecification(ctx, msg.ContractSpecificationId)
+	if !found {
+		return nil, fmt.Errorf("contract specification not found with id %s", msg.ContractSpecificationId)
+	}
+	if err := k.ValidateAllOwnersAreSigners(contractSpec.OwnerAddresses, msg.Signers); err != nil {
```
Comment (iramiller) [inline x/metadata/keeper/msg_server.go:381]: the object being updated is the scope specification ... not the contract specification.  The contract specification owners are not required to sign as their object is not changing.

File: x/metadata/keeper/msg_server.go [hunk 6]
```diff
@@ -368,6 +368,83 @@ func (k msgServer) DeleteContractSpecification(
...
+	goCtx context.Context,
+	msg *types.MsgDeleteContractSpecFromScopeSpecRequest,
+) (*types.MsgDeleteContractSpecFromScopeSpecResponse, error) {
+	defer telemetry.MeasureSince(time.Now(), types.ModuleName, "tx", "DeleteContractSpecFromScopeSpec")
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	contractSpec, found := k.GetContractSpecification(ctx, msg.ContractSpecificationId)
+	if !found {
+		return nil, fmt.Errorf("contract specification not found with id %s", msg.ContractSpecificationId)
+	}
+	if err := k.ValidateAllOwnersAreSigners(contractSpec.OwnerAddresses, msg.Signers); err != nil {
```
Comment (iramiller) [inline x/metadata/keeper/msg_server.go:415]: contract spec is not being modified here so its owners are not required to sign this update.

File: x/metadata/types/msg.go [hunk 7]
```diff
@@ -871,6 +875,86 @@ func (msg MsgDeleteContractSpecificationRequest) ValidateBasic() error {
...
+func (msg MsgAddContractSpecToScopeSpecRequest) GetSignBytes() []byte {
+	return sdk.MustSortJSON(ModuleCdc.MustMarshalJSON(&msg))
+}
+
+// ValidateBasic performs a quick validity check
+func (msg *MsgAddContractSpecToScopeSpecRequest) ValidateBasic() error {
+	if len(msg.Signers) < 1 {
+		return fmt.Errorf("at least one signer is required")
+	}
+	return nil
```
Comment (iramiller) [inline x/metadata/types/msg.go:915]: consider adding validation of the metadata address format and type here.

File: x/metadata/types/msg.go [hunk 8]
```diff
@@ -871,6 +875,86 @@ func (msg MsgDeleteContractSpecificationRequest) ValidateBasic() error {
...
+func (msg MsgDeleteContractSpecFromScopeSpecRequest) GetSignBytes() []byte {
+	return sdk.MustSortJSON(ModuleCdc.MustMarshalJSON(&msg))
+}
+
+// ValidateBasic performs a quick validity check
+func (msg MsgDeleteContractSpecFromScopeSpecRequest) ValidateBasic() error {
+	if len(msg.Signers) < 1 {
+		return fmt.Errorf("at least one signer is required")
+	}
+	return nil
```
Comment (iramiller) [inline x/metadata/types/msg.go:955]: consider adding validation of the metadata address format and type here

**Review by iramiller — CHANGES_REQUESTED** [review]
A few comments.  Also see the signer requirement comments.  It is not required to sign with contract spec owners as the contract spec is not being modified here.

**Comment by channa-figure** [comment]
@iramiller thanks for the detailed review.  I got implemented those changes and wrote/updated tests.

**Review by iramiller — APPROVED** [review]
Those cli commands are pretty ugly/long but then again I don't really see them being used very often.

File: x/metadata/client/cli/tx.go [hunk 9]
```diff
@@ -1033,6 +1076,46 @@ func RemoveContractSpecificationCmd() *cobra.Command {
 	return cmd
 }
 
+// RemoveContractSpecFromScopeSpecCmd removes a contract spec from scope spec command
+func RemoveContractSpecFromScopeSpecCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "remove-contract-spec-to-scope-spec contract-specification-id scope-specification-id",
```
[the remark this answers is not in the document]
Comment (channa-figure) [reply x/metadata/client/cli/tx.go]: 🤦 Good catch!

**Comment by channa-figure** [comment]
I agree that the command names are long and ugly.  I stared at them for a while wondering how I could shorten them and have them make sense.  However, I came to the conclusion that they probably won't be used too often.  I added an alias to both the commands to shorten what needs to be typed.

**Review by iramiller — APPROVED** [review]
LGTM.

**Pull request merged**